### ISSUE TITLE
dcc.Store and encrypted props

### ISSUE THREAD

**Issue by nicolaskruchten** [plotly/dash#262]
We should have a better/safer way of storing data on the client than CSS-invisible `html.Div` components, for sensitive data that not all users should see/have access to. I propose a new `dcc.Store` component which wouldn't render in the DOM and which would transparently encrypt/decrypt its `data-*` props on the server.

#254 was a first cut at implementing this in `dash.py` and some design discussion is available there. It seems like the way forward here is to make the base component class extensible with server-side `transform`/`inverse_transform` extension points.

@chriddyp is there an existing issue that captures other needs/thoughts around the base component extension idea?

**Comment by valentijnnieman** [comment]
@nicolaskruchten Can this be closed now that dash-core-components has a new `Store` component?

**Comment by nicolaskruchten** [comment]
No, because the data stored there is still unencrypted :)

**Comment by supernyv** [comment]
Hi guys,

Nice to see this discussion. This is exactly what I am looking for, a dcc.Store that has the capability to encrypt and decrypt its data, and it will probably help a lot of developers

**Comment by gvwilson** [comment]
@AnnMarieW  will `dcc.Store` do encryption these days?

**Comment by AnnMarieW** [comment]
No, this feature is still not available, but I could see how it could be useful.

**Comment by gvwilson** [comment]
cc @ndrezn what do you think?

**Comment by ndrezn** [comment]
TLDR: This should be possible today, and I'm not sure if baking it into the component would even work... because as @BSd3v says we always need to store the encryption key server side somewhere. But if we think it's something to consider we should open a new "feature request ticket" (this ticket IMHO should be closed because `dcc.Store` exists 🙂).

---

I think it should be pretty straightforward to implement this today, by encrypting the data before sending it to `dcc.Store` using a key available on the server side, and decrypting using the same key on the way in. Like @BSd3v mentions this wouldn't play nicely with client side callbacks, because of course you can't send the key to the client.

I think it's probably out of scope for `dcc.Store` to support this because it means we need to support encryption types, possibly add encryption libraries as dependencies... this feels like a good "it works in Python, so it works in Dash" feature IMO.

Maybe I'm missing something, but why would you want do to this...?

The main case I can think of is sending all data to a store encrypted and then using that store across your app to show only subsets of your data. But if the data is that sensitive something like a database would probably be more secure/easier to manage (and of course scale better).

Here's a small app demonstrating, where clicking "Submit" encrypts the data and sends it to `dcc.Store`, and updates to `dcc.Store` update the `html.Div` to show what's going on. You more likely would not want the second callback...

This example keeps the key alive in memory. An environment variable would probably also work fine as long as you never `return` the key from a callback.

```console
pip install dash cryptography
```
Then:
```python
from dash import dcc, html, Input, Output, State, callback, Dash, no_update
from cryptography.fernet import Fernet
import base64

# Generate a key for encryption and decryption
key = Fernet.generate_key()
cipher_suite = Fernet(key)

app = Dash(__name__)

app.layout = html.Div(
    [
        dcc.Input(
            id="input-data",
            type="text",
            placeholder="Enter data to encrypt",
            style={"marginBottom": "10px"},
        ),
        html.Button("Submit", id="submit-button", n_clicks=0),
        dcc.Store(id="encrypted-store"),
        html.Div(id="output"),
    ]
)


@callback(
    Output("encrypted-store", "data"),
    Input("submit-button", "n_clicks"),
    State("input-data", "value"),
)
def encrypt_data(n_clicks, value):
    if n_clicks > 0 and value:
        encrypted_data = cipher_suite.encrypt(value.encode())
        # Encode to base64 to store in dcc.Store
        encoded_data = base64.b64encode(encrypted_data).decode()
        return encoded_data
    return no_update


@callback(Output("output", "children"), Input("encrypted-store", "data"))
def decrypt_data(encoded_data):
    if encoded_data:
        # Decode from base64 and decrypt
        encrypted_data = base64.b64decode(encoded_data)
        decrypted_data = cipher_suite.decrypt(encrypted_data).decode()
        return f"Encrypted data: {encoded_data}\n Decrypted data: {decrypted_data}"
    return "No data stored"


if __name__ == "__main__":
    app.run_server(debug=True)

```

**Comment by nicolaskruchten** [comment]
> not totally sure whether automatic encryption of dcc.Store props is super useful or significantly simplified by having it be automatic.

The idea here was to provide a first-class/recommended way of doing things that doesn't require users to copy-paste boilerplate around.

**Comment by ndrezn** [comment]
For my own reference, @chriddyp puts the use case well in: https://community.plotly.com/t/writing-secure-dash-apps-community-thread/54619
> 2. `dcc.Store` Data is accessible, even if it’s not visible
>
> All of the data that is returned in a callback can be accessed by the current user of the app, even if that data isn’t necessarily visible.
>
> So, be careful not to store secret or sensitive data in dcc.Store or other callbacks if you don’t intend all users to be able to access the data.
>
> For example, if you have a dataframe that has several unused but sensitive columns (e.g. data about a particular client), then remove those columns before saving it in a dcc.Store. Even if your app doesn’t use those columns, the data is still accessible in the browser session.

**Comment by BSd3v** [comment]
What if Plotly added support for Server Side stores, much like `dash-extensions`. Then you wouldnt have to work about keeping things private.

Also, another thing to point out for a dcc.Store and encrypting on the server, you wouldnt be able to utilize the data on the browser via a clientside callback. So, the point of even having the data go to the client doesnt make much sense.

This of course would be different if for some reason you wanted to send the data via an api call by the client...
